google/cloud/aiplatform/_featurestores/featurestore.py

```diff
@@ -0,0 +1,322 @@
+                fixed_node_count=fixed_node_count
+            ),
+        )
+
+        _LOGGER.log_action_start_against_resource(
+            "Updating", "featurestore", self,
+        )
+
+        update_featurestore_lro = self.api_client.update_featurestore(
+            featurestore=gapic_featurestore,
+            update_mask=update_mask,
+            metadata=request_metadata,
+        )
+
+        _LOGGER.log_action_started_against_resource_with_lro(
+            "Update", "featurestore", self.__class__, update_featurestore_lro
+        )
+
+        update_featurestore_lro.result()
+
+        _LOGGER.log_action_completed_against_resource("featurestore", "updated", self)
+
+        return self
+
+    def list_entity_types(
+        self, filter: Optional[str] = None, order_by: Optional[str] = None,
+    ) -> List["_featurestores.EntityType"]:
+        """Lists existing managed entityType resources in this Featurestore.
+
+        Example Usage:
+
+            my_featurestore = aiplatform.Featurestore(
+                featurestore_name='my_featurestore_id',
+            )
+            my_featurestore.list_entity_types()
+
+        Args:
+            filter (str):
+                Optional. Lists the EntityTypes that match the filter expression. The
+                following filters are supported:
+
+                -  ``create_time``: Supports ``=``, ``!=``, ``<``, ``>``,
+                   ``>=``, and ``<=`` comparisons. Values must be in RFC
+                   3339 format.
+                -  ``update_time``: Supports ``=``, ``!=``, ``<``, ``>``,
+                   ``>=``, and ``<=`` comparisons. Values must be in RFC
+                   3339 format.
+                -  ``labels``: Supports key-value equality as well as key
+                   presence.
+
+                Examples:
+
+                -  ``create_time > \"2020-01-31T15:30:00.000000Z\" OR update_time > \"2020-01-31T15:30:00.000000Z\"``
+                   --> EntityTypes created or updated after
+                   2020-01-31T15:30:00.000000Z.
+                -  ``labels.active = yes AND labels.env = prod`` -->
+                   EntityTypes having both (active: yes) and (env: prod)
+                   labels.
+                -  ``labels.env: *`` --> Any EntityType which has a label
+                   with 'env' as the key.
+            order_by (str):
+                Optional. A comma-separated list of fields to order by, sorted in
+                ascending order. Use "desc" after a field name for
+                descending.
+
+                Supported fields:
+
+                -  ``entity_type_id``
+                -  ``create_time``
+                -  ``update_time``
+
+        Returns:
+            List[EntityTypes] - A list of managed entityType resource objects.
+        """
+        return _featurestores.EntityType.list(
+            featurestore_name=self.resource_name, filter=filter, order_by=order_by,
+        )
+
+    @base.optional_sync()
```

Comment:
Just realized that this is decorated. I think this is new usage pattern for the SDK and there are three options on usage here to consider. This is worth discussion because this pattern should be used throughout.

**1**:  if this method is decorated then the usage is:

```Python
 feature_store.delete_entity_types(..., sync=False)
 feature_store.wait()  # will throw if any of the deletions had issues
 ```

In this case there is no need to pass `sync` to the underlying deletions and instead we can default to asynchronous to make deletions concurrent. Also note that the method needs to be changed to wait on all entity_type deletes not just the last one as it is currently handled.

```Python 
entity_types = []
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=False)
  entity_types.append(entity_type)

for entity_type in entity_types:
  entity_type.wait()
```

**2**: other option is to pass `sync` directly to the underlying delete calls but not associate them with the FeatureStore:

```Python
 feature_store.delete_entity_types(..., sync=False)
 feature_store.wait()  # will NOT throw if any of the deletions had issues
```

In this case this method should not be decorated with `base.optional_sync`. There is no need to call wait here and the implementation would look like:

```Python 
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=sync)
``` 

**3**: is similar to the second option but instead we can return the list of deleted entity types so the user can wait on each one. So usage look like:

```Python
deleted_entity_types: List[EntityType] = feature_store.delete_entity_types(..., sync=False)
feature_store.wait()  # will NOT throw if any of the deletions had issues

for entity_type in deleted_entity_types:
  entity_type.wait()
```

The implementation would look similar to (2) but we compose the list and return it.

```Python 
entity_types = []
for entity_type_id in entity_type_ids:
  entity_type = self.get_entity_type(entity_type_id=entity_type_id)
  entity_type.delete(sync=sync)
  entity_types.append(entity_type)
return entity_types
``` 

Also to unblock this PR, feel free to remove the async support for now and follow up. Up to you.

Reply:
Thanks for the options. I would vote for option 1 over 2 and 3. The main reason is when we delete child resources from the parent resource, I do want to provide sync option to prevent user to retrieve the delete pending child from the parent, i.e. Featurestore.get_entity_type() when Featurestore.delete_entity_types(). So add a Featurestore.wait() during deleting kind of make sense, and if any of the deletion had issue, it is captured.

Reply:
SGTM